### PR TITLE
Refactor some of the cluster install clients to be testable

### PR DESCRIPTION
**Pull request by hawkowl**
BASED OFF https://github.com/Azure/ARO-RP/pull/1129

### Which issue this PR addresses:

Part of https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/6118707/

### What this PR does / why we need it:

This refactors some of the Azure clients to be injected as factories into the cluster installer to allow for testing/a bit cleaner abstraction for some places.

### Test plan for issue:

Comes with some tests, and enables more tests I have locally.

### Is there any documentation that needs to be updated for this PR?

N/A

### CODE REVIEW

**Comment by jim-minter** [comment]
@hawkowl I've rebased this for you, hopefully correctly.

File: pkg/util/blob/blob.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
+package blob
```
Comment (jim-minter) [inline pkg/util/blob/blob.go:1]: This file is joining two things together which I'd rather keep separate.
1) The pkg/cluster-specific code by which we get access to cluster blobs (NewBlobServiceForCluster)
2) Modification of the blob storage API for supposed better usability/testability.

WRT 1), I think it makes sense to keep this code in pkg/cluster rather than move it to pkg/util.

WRT 2), if/when we add egress lockdown code it is likely we will want to use storage blobs and queues more.  For that reason I would rather avoid adding smarts to the util storage API but rather keep very true to the original like we've done for ARM clients in pkg/util/azureclient up to now.  It will make it easier to extend the wrapper.  PTAL at https://github.com/openshift/openshift-azure/blob/master/pkg/util/azureclient/storage/storage.go (excluding the retry code which should probably not have gone in there) for an example of the approach I'd prefer to take if possible.

Following 2), there should be fewer/no changes to pkg/cluster/graph.go, etc., for example, as we're not changing the native API.

**Review by jim-minter — CHANGES_REQUESTED** [review]

File: pkg/util/blob/blob.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	return err
+}
+
+func (b *blobService) DeleteContainerIfExists(container string) error {
+	cont := b.c.GetContainerReference(container)
+	_, err := cont.DeleteIfExists(&azstorage.DeleteContainerOptions{})
+	return err
+}
+
+type fakeBlob struct {
```
Comment (jim-minter) [inline pkg/util/blob/blob.go:99]: probably needs to live somewhere else - separate file, maybe separate pkg under /test ?

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -74,6 +77,10 @@ type manager struct {
 	samplescli    samplesclient.Interface
 	securitycli   securityclient.Interface
 	arocli        aroclient.AroV1alpha1Interface
+
+	newDynamicValidator   func(*logrus.Entry, env.Interface, *api.OpenShiftCluster, *api.SubscriptionDocument, refreshable.Authorizer) validate.OpenShiftClusterDynamicValidator
+	getClusterSPObjectID  func(context.Context, env.Interface, *logrus.Entry, *api.OpenShiftClusterDocument) (string, error)
+	getBlobServiceBackend func(context.Context, storage.AccountsClient, *api.OpenShiftCluster, mgmtstorage.Permissions, mgmtstorage.SignedResourceTypes) (blob.BlobService, error)
```
Comment (ihcsim) [inline pkg/cluster/cluster.go:83]: What's the motivation behind these changes? Is the plan also to mock these fields for unit testing 
 purposes, like `newDynamicValidator`? AFAICT, they are only instantiated in `pkg/cluster/cluster.go`, and not mocked in any existing tests:

```
$ grep -rniI "getblobservicebackend:\|getclusterspobjectid:" * 
pkg/cluster/cluster.go:134:             getClusterSPObjectID:  clusterSPObjectID,
pkg/cluster/cluster.go:135:             getBlobServiceBackend: blob.NewBlobServiceForCluster,
```

Just wondering if these changes are necessary.

File: pkg/cluster/generateconfig_test.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+			}
+
+			installConfig, platformCreds, image, err := c.generateInstallConfig(ctx)
+
+			if tt.wantErr == "" {
+				if err != nil {
+					t.Error(err)
+				}
+				if installConfig == nil {
+					t.Error(installConfig)
```
Comment (ihcsim) [inline pkg/cluster/generateconfig_test.go:180]: Might want to consider exiting here with `t.Fatal()`. Otherwise, if `installConfig == nil`, the test will panic at line 201:

```
go test ./pkg/cluster/... -run="TestBackendTry"         
--- FAIL: TestBackendTry (0.00s)                                                                                                                  
    --- FAIL: TestBackendTry/valid_document_generates_valid_installconfig (0.00s)                                                                 
        generateconfig_test.go:181: <nil>                                                                                                         
panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                               
        panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                       
        panic: runtime error: invalid memory address or nil pointer dereference                                                                   
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2d419b9]                                                                           
                                                                                                                                                  
goroutine 83 [running]:                                                                                                                           
testing.tRunner.func1.1(0x3350240, 0x5f5c2c0)                                                                                                     
        /usr/local/go/src/testing/testing.go:999 +0x319                                                                                           
testing.tRunner.func1(0xc00079e7e0)                                                                                                               
        /usr/local/go/src/testing/testing.go:1002 +0x402                                                                                          
panic(0x3350240, 0x5f5c2c0)                                                                                                                       
        /usr/local/go/src/runtime/panic.go:969 +0x166                                                                                             
github.com/golang/mock/gomock.(*Controller).Finish(0xc0007a8540)
        /home/isim/workspace/go/src/github.com/Azure/ARO-RP/vendor/github.com/golang/mock/gomock/controller.go:246 +0x27a
panic(0x3350240, 0x5f5c2c0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/Azure/ARO-RP/pkg/cluster.TestBackendTry.func3(0x0, 0xc0008cdd60, 0x1, 0x1)
        /home/isim/workspace/go/src/github.com/Azure/ARO-RP/pkg/cluster/generateconfig_test.go:102 +0x29
github.com/Azure/ARO-RP/pkg/cluster.TestBackendTry.func8(0xc00079e7e0)
        /home/isim/workspace/go/src/github.com/Azure/ARO-RP/pkg/cluster/generateconfig_test.go:201 +0x871
testing.tRunner(0xc00079e7e0, 0xc000212b80)
        /usr/local/go/src/testing/testing.go:1050 +0xdc
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1095 +0x28b
FAIL    github.com/Azure/ARO-RP/pkg/cluster     0.022s
```

**Comment by mjudeikis** [comment]
I think this is going to the right direction. We already started going towards `Manager` interfaces with DNS, ACRToken.
Why not:
```
type Manager interface {
	GetToken(ctx context.Context) (*adal.ServicePrincipalToken, error)
	GetServicePrincipalID(ctx context.Context) (string, error)
}

type manager struct {
	log             *logrus.Entry
	env             env.Interface
	oc              *api.OpenShiftCluster
	subscriptionDoc *api.SubscriptionDocument
}

func NewManager(log *logrus.Entry, env env.Interface, oc *api.OpenShiftCluster, subscriptionDoc *api.SubscriptionDocument) Manager {
	return &manager{
		log:             log,
		env:             env,
		oc:              oc,
		subscriptionDoc: subscriptionDoc,
	}
}
```
Which would be part of the cluster main struct. Struct already operates on single document at the time. 
In addition this keeps mock interfaces in the same state as all other.

In addition one big blocker to test this would be `m.deployARMTemplate(ctx, resourceGroup, "storage", t, nil)` as it will be used widely by upgade path too. So it might need to get its own "manager"

**Comment by hawkowl** [comment]
Gonna give this one a viking funeral until I can come back to it.
